tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,433 @@
+                            const Index* indices, T* selected) {
+  CUDA_AXIS_KERNEL_LOOP(idx, config.virtual_thread_count, X) {
+    selected[idx] = original[indices[idx]];
+  }
+}
+
+template <typename P1, typename P2, typename Index>
+__global__ void PairedSelect(const CudaLaunchConfig config, const P1* p1,
+                             const P2* p2, const Index* indices,
+                             P1* selected_p1, P2* selected_p2) {
+  for (auto idx : CudaGridRangeX(config.virtual_thread_count)) {
+    selected_p1[idx] = p1[indices[idx]];
+    selected_p2[idx] = p2[indices[idx]];
+  }
+}
+
+template <typename T>
+__global__ void Iota(const CudaLaunchConfig config, const T offset,
+                     T* to_fill) {
+  for (int idx : CudaGridRangeX(config.virtual_thread_count)) {
+    to_fill[idx] = static_cast<T>(idx) + offset;
+  }
+}
+
+tensorflow::Status nms_gpu(const float* d_desc_sorted_boxes_float_ptr,
+                           const int N, const float thresh,
+                           int* d_keep_sorted_list, int* h_nkeep,
+                           int* dev_delete_mask, int* host_delete_mask,
+                           OpKernelContext* context, bool flip_boxes) {
+  // d_desc_sorted_boxes_float_ptr is a pointer
+  //    to device memory float array containing the box corners for N boxes.
+  // N is number of boxes.
+  // threshold is the iou threshold for elimination
+  // d_keep_sorted_list is a device pointer to int array containing sorted
+  //    indices of the boxes to keep h_nkeep is the pointer to number of
+  //    elements to keep for the host.
+  // h_nkeep is a host pointer for returning number of items to keep
+  // flip_boxes flag reorders the boxes use lower left and upper right corners
+  // if they are given in mixed format. Making sure we respect the __align(16)__
+  // we promised to the compiler
+  auto iptr = reinterpret_cast<std::uintptr_t>(d_desc_sorted_boxes_float_ptr);
+  CHECK_EQ((iptr & 15), 0);
+
+  // The next kernel expects squares
+
+  const int mask_ld = (N + NMS_BOXES_PER_THREAD - 1) / NMS_BOXES_PER_THREAD;
+  const Box* d_desc_sorted_boxes =
+      reinterpret_cast<const Box*>(d_desc_sorted_boxes_float_ptr);
+  auto stream_exec = context->op_device_context()->stream();
+  auto device = context->eigen_gpu_device();
+  dim3 grid, block;
+  int block_size = (N + NMS_BLOCK_DIM - 1) / NMS_BLOCK_DIM;
+  block_size = std::max(std::min(block_size, NMS_BLOCK_DIM_MAX), 1);
+  grid.x = block_size;
+  grid.y = block_size;
+  block.x = NMS_BLOCK_DIM;
+  block.y = NMS_BLOCK_DIM;
+  NMSKernel<<<grid, block, 0, device.stream()>>>(
+      d_desc_sorted_boxes, N, thresh, mask_ld, dev_delete_mask, flip_boxes);
+  CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
+  int nto_copy = std::min(NMS_CHUNK_SIZE, N);
+  cudaEvent_t copy_done;
+  cudaEventCreate(&copy_done);
```

Comment:
Use stream_executor::Event

Reply:
@csigg I couldn't be able to find any examples of using stream_executor::Event in a similar fashion elsewhere in kernels. Even then, I don't think it is possible to implement the logic by stream_executor::Event since there is no mechanism equivalent to cudaEventSynchronize() implemented in the framework. I can try to spin on event::poll but that would be quite inefficient and would probably hinder the rest of the framework as well due to acquired locks. I would have preferred to use ThenExecute() chaining these but it would require all NMS ops to be converted to AsyncOps as well as a proper threadpool on event manager. Currently all events are executed on single thread and doing work there would block the event infrastructure. I can spawn the work on cpu device thread pool on the event callback but I am not sure if this level of complexity is justified.
How would you propose I would use stream_executor::Event, it is possible that I am missing something obvious.